### PR TITLE
Add some OAS3 param serialization 

### PR DESCRIPTION
**Pull request by iainbryson**
Beginning support:

* Support `schema: { type: object }` in `param`
* Support `style: :form` with `explode` = `true` or `false`
* Support `style: :deepObject`

for issue #271

### CODE REVIEW

**Review by samuelgiles — APPROVED** [review]
Nice 👏 I just bumped into the exact same issue and was pleasantly surprised to find not only was the OAS 3.0 type -> schema.type change in here but the deepObject/Object/explode changes too 👍

File: rswag-specs/lib/rswag/specs/request_factory.rb [hunk 1]
```diff
@@ -118,6 +118,21 @@ def add_path(request, metadata, swagger_doc, parameters, example)
 
       def build_query_string_part(param, value)
         name = param[:name]
+
+        # OAS 3: https://swagger.io/docs/specification/serialization/
+        if param.has_key? :schema
+          return "#{name}=#{value}" unless param[:schema][:type].to_sym == :object
+
+          style = param[:style] || :form
```
Comment (samuelgiles) [inline rswag-specs/lib/rswag/specs/request_factory.rb]: Thinking about it, as per the pattern set by `param[:schema][:type].to_sym` I think it'd be nice to allow `style: 'deepObject'` to work here too by doing something like:

```suggestion
          style = (param[:style] || :form).to_sym
```

**Comment by iainbryson** [comment]
Thanks @samuelgiles, that's a great suggestion

**Comment by samuelgiles** [comment]
@iainbryson I forked your changes and switched to using ActiveSupport's `#to_query` as opposed to manually building up the deep object, this in turn brings support for nested objects (and some URI safety re: the namespacing):

https://github.com/samuelgiles/rswag/commit/ae43cb0abd6ecaf577293208da0ca6b4cacb9a93

What do you think about cherry picking that commit into this PR?

**Comment by iainbryson** [comment]
Hey @samuelgiles I've cherry-picked your commit and followed the same theme of using the AR `to_query` on the `:form` style with `:explode`, and url-encoded the names for the non-`:explode` version.

**Review by samuelgiles — APPROVED** [review]
Awesome, thanks for that 👍

**Comment by iainbryson** [comment]
@samuelgiles what happens next?  What's the protocol for getting this merged.

**Comment by BookOfGreg** [comment]
@iainbryson the current trick is every few weeks/months depending on how mental work is I do a sweep for PRs that are documented + tested, I check it looks sensible and follows spec, then I go on a merging spree late at night and push out a version. 😄 

I've granted a couple other people writer access this year so there's now 4 of us able to commit/release but I don't see them around too often.

Maintainers wanted!

File: rswag-specs/lib/rswag/specs/request_factory.rb [hunk 2]
```diff
@@ -118,6 +118,26 @@ def add_path(request, metadata, swagger_doc, parameters, example)
 
       def build_query_string_part(param, value)
         name = param[:name]
+
+        # OAS 3: https://swagger.io/docs/specification/serialization/
+        if param.has_key?(:schema)
```
Comment (akofink) [inline rswag-specs/lib/rswag/specs/request_factory.rb]: We should check the doc_version here instead:

```suggestion
        if doc_version(swagger_doc).start_with?('3')
```

File: rswag-specs/spec/rswag/specs/request_factory_spec.rb [hunk 3]
```diff
@@ -103,6 +103,64 @@ module Specs
...
+          before do
+            metadata[:operation][:parameters] = [
+              {
+                name: 'things', in: :query,
+                style: style,
+                explode: explode,
+                schema: { type: :object, additionalProperties: { type: :string } }
+              }
+            ]
+            allow(example).to receive(:things).and_return(things)
```
Comment (akofink) [inline rswag-specs/spec/rswag/specs/request_factory_spec.rb:119]: These are all OpenAPI 3 tests, so we should set the doc_version:

```suggestion
            allow(example).to receive(:things).and_return(things)
            let(:swagger_doc) { { swagger: '3.0' } }
```

**Review by akofink — COMMENTED** [review]
Awesome! I have some comments inline.

**Comment by iainbryson** [comment]
@akofink Sorry for the long delay -- work :(. I've taken the feedback and fixed up the tests & code to fix some issues resulting from it.  All tests pass now.

**Review by akofink — APPROVED** [review]
Looks great, thanks! :100:

**Comment by kovyrin** [comment]
Is there any chance this PR will be merged any time soon?

**Comment by ngouy** [comment]
Hello there (cc @BookOfGreg @iainbryson). First thanks a lot for that, we needed this!

Though I'm not quite sure, It seems like it breaks arrays params. Indeed given this piece of code
```rb
        # OAS 3: https://swagger.io/docs/specification/serialization/
        if swagger_doc && doc_version(swagger_doc).start_with?('3') && param[:schema]
          return "#{name}=#{value}" unless param[:schema][:type].to_sym == :object
```

As long as you find the 3.0 notation, you completely obliterate the second part (the old code) that was handling arrays properly.

Is it a bug? Or is it how 3.0 is meant to work

I am pretty much sure specs woul broke if you add the `let(:swagger_doc) { { swagger: '3.0' } }` to the array section

*edit* now I realize it's a mix of both, 3.0 evolved, but there is a bug because we miss proper implementation for arrays

**Review by ebakep — COMMENTED** [review]
🔥

**Pull request merged**